Function pointer non-det initialisation in goto-harness [depends-on: #5171]

NlightNFotis opened the pull request:
This is extending the work done in #5171 of which the first commit of this PR belongs to.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -797,3 +801,54 @@ bool recursive_initializationt::needs_freeing(const exprt &expr) const
+  body.add(
+    code_assignt{function_pointer_selector,
+                 side_effect_expr_nondett{function_pointer_selector.type()}});
+  auto function_pointer_index = 0;
```
`std::size_t` .. this way the deduced type is `int`.

NlightNFotis replied:
Done.

xbauch reviewed (approved):
Looks good.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 825029acf7b833dcff85c26eab549dffa46e6f07).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/135629595

NlightNFotis committed:
Apply miscellaneous refactoring and fixes to goto-harness.

Add error checking (check that the entry function is in the symbol table
before the code generation, and make sure that it's present correctly)
make sure that doc comments are capitalised, change the signatures of
functions to make them stricter in the arguments they accept.

NlightNFotis committed:
Add support for function pointers to goto-harness

Also add support for having multiple constructors for the
same type with different behaviours and different signatures.
We are doing this because we need some types to be sometimes
nullable and sometimes not, and for example, for arrays we
sometimes need to pass a size parameter and sometimes not.

NlightNFotis committed:
Map function pointer parameter to function pointer argument

Add mapping from function pointer argument names to
local variable names of the harness so users can specify
properties of arguments in terms of the names of the
arguments rather than having to rely on internal
implementation details of the function harness.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 60398e78d5a429968028e0357b887a49d1aa1964).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136133258

NlightNFotis committed:
Add more tests.

Add a test to make sure that we can distinguish between
a parameter and a global function pointer variable with
the same name and a test to make sure that we get
different type constructors for the same type but with
extra attributes.

NlightNFotis committed:
Move function-pointer-can-be-null option handling.

Move it to recursive_initialization_configt because
it's a common option.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 64a62d4986957f945dc1eaf129c77fde8e205a3f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136162388

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dfadbc87b35ca3d1e151499daba2cd314f40e62b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136177150

xbauch commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -12,6 +12,7 @@ Author: Diffblue Ltd.
 #include <util/arith_tools.h>
 #include <util/c_types.h>
 #include <util/exception_utils.h>
+#include <util/prefix.h>
```
do you still need this? (edit: you do, but somewhere in the 4th commit; no need to reorder because of this.)

hannes-steffenhagen-diffblue replied:
I think it's worth fixing that.

xbauch commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -264,11 +275,12 @@ irep_idt recursive_initializationt::build_constructor(const exprt &expr)
   goto_model.goto_functions.function_map[function_symbol.name].type =
     to_code_type(function_symbol.type);
-  return type_constructor_names[type];
+  return type_constructor_names.at(key);
```
Is there a reason why this access is checked but the above are not?

hannes-steffenhagen-diffblue replied:
Not really, tbh we'd considered refactoring this entire section to avoid multiple lookup of the same key. We only added the `.at` because at some point we were hitting a case where keys were missing from the map.

xbauch commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -203,17 +208,23 @@ irep_idt recursive_initializationt::build_constructor(const exprt &expr)
   {
     expr_name = to_symbol_expr(expr).get_identifier();
+    is_nullable = initialization_config.potential_null_function_pointers.count(
+      expr_name.value());
```
Maybe it's idiomatic, but I would prefer `count(..) != 0`

hannes-steffenhagen-diffblue replied:
I agree that'd be easier to read.

xbauch commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -335,6 +347,43 @@ void function_call_harness_generatort::validate_options(
+    }
+
+    const auto &function_pointer_type =
+      ns.follow(function_pointer_symbol_pointer->type);
```
Why do we need to follow the type?

hannes-steffenhagen-diffblue replied:
Symbol types aren't resolved at this point (i.e. in the symbol table). They can't be, otherwise we couldn't correctly print typedefs etc in dump-c.

xbauch commented on src/goto-harness/common_harness_generator_options.h:
```diff
@@ -14,13 +14,16 @@ Author: Diffblue Ltd.
   "(" COMMON_HARNESS_GENERATOR_MAX_NONDET_TREE_DEPTH_OPT "):"                  \
   "(" COMMON_HARNESS_GENERATOR_MIN_ARRAY_SIZE_OPT "):"                         \
   "(" COMMON_HARNESS_GENERATOR_MAX_ARRAY_SIZE_OPT "):"                         \
+  "(" COMMON_HARNESS_GENERATOR_FUNCTION_POINTER_CAN_BE_NULL_OPT "):"           \
```
Please add a help description below as well.

xbauch commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -74,6 +74,17 @@ bool recursive_initialization_configt::handle_option(
+    std::transform(
+      values.begin(),
+      values.end(),
+      std::inserter(
```
For the Windows build, you are going to have to `#include <iterator>`

hannes-steffenhagen-diffblue replied:
Thanks, I figured it'd be something like that!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 272477c724339d4cfe65e58d4ba64816f4a6061b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136582105

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -72,10 +72,10 @@ struct function_call_harness_generatort::implt
   /// Update the goto-model with the new harness function.
   void add_harness_function_to_goto_model(code_blockt function_body);
-  /// declare local variables for each of the parameters of the entry function
+  /// Declare local variables for each of the parameters of the entry function
```
44509e46ce8d4092a806d3d7bb9c767555ce1932 are good changes, but might be better off in a separate PR (although it’d be annoying to separate out this late I know)

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -799,3 +811,57 @@ bool recursive_initializationt::needs_freeing(const exprt &expr) const
+
+  for(const auto &sym : goto_model.get_symbol_table())
+  {
+    if(sym.second.type == function_type)
```
We should probably make this work with the function pointer restriction table stuff we have some time.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -57,7 +58,29 @@ class recursive_initializationt
+    typet constructor_type;
+    bool is_nullable;
+    bool has_size_parameter;
+    bool operator<(const constructor_keyt &other) const
```
With C++20 we could just default these ;)

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -165,19 +188,29 @@ class recursive_initializationt
+  /// \return the body of the constructor
+  code_blockt build_function_pointer_constructor(
+    const symbol_exprt &result,
+    bool is_nullable);
```
I’m not extremely excited about having to pass this `is_nullable` argument down all the time, but I can’t think of anything super great either.

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -81,6 +81,16 @@ struct function_call_harness_generatort::implt
+  /// this specification from that for a global variable of the same name,
+  /// we prepend the name of the function to the parameter name. However,
+  /// what is actually being initialised in the implementation is not the
+  /// parameter itself, but a correspond function argument (local variable
```
correspond(ing)
